### PR TITLE
tests: version handling: streamline to single version provider

### PR DESCRIPTION
**Pull request by nrainer-materialize**
This PR streamlines the provider for used versions and predecessor versions. In particular, it removes / replaces `VersionsFromGit` and `VersionsFromDocs`. It is a follow-up to https://github.com/MaterializeInc/materialize/pull/23145.

### Status
Ready for review now. Result of nightly is good.

### Important notes
* The order of versions provided by some utility methods changed to newest-first, e.g., `get_published_minor_mz_versions()` returns the latest version as first element.
* There is a check / filter that makes sure that the docker image of recent versions exist. This is done by pulling the image. As performance optimization, this is not done for versions older than a certain specified version.

### Commits
Commit-wise review might be easier due to some moved code.

### Nightly builds
https://buildkite.com/materialize/nightlies/builds?branch=nrainer-materialize%3Aci%2Fversion-resolution

### CODE REVIEW

**Comment by nrainer-materialize** [comment]
Issue in mkpipeline to be addressed:

```
Trying to pull image: materialize/materialized:v0.78.10
Trying to pull image: materialize/materialized:v0.78.9
Trying to pull image: materialize/materialized:v0.78.8
Trying to pull image: materialize/materialized:v0.78.7
Failed to pull image: materialize/materialized:v0.78.7
Trying to pull image: materialize/materialized:v0.78.6
Failed to pull image: materialize/materialized:v0.78.6
Trying to pull image: materialize/materialized:v0.78.5
Failed to pull image: materialize/materialized:v0.78.5
Trying to pull image: materialize/materialized:v0.78.4
Trying to pull image: materialize/materialized:v0.78.3
Trying to pull image: materialize/materialized:v0.78.2
Failed to pull image: materialize/materialized:v0.78.2
Trying to pull image: materialize/materialized:v0.78.1
Failed to pull image: materialize/materialized:v0.78.1
Trying to pull image: materialize/materialized:v0.78.0
Failed to pull image: materialize/materialized:v0.78.0
Trying to pull image: materialize/materialized:v0.77.1
Trying to pull image: materialize/materialized:v0.77.0
Trying to pull image: materialize/materialized:v0.78.10
Trying to pull image: materialize/materialized:v0.77.1
```

**Comment by def-** [comment]
>  platform-checks-materialized-1      | environmentd: 2023-11-28T17:27:43.229846Z  WARN environmentd::run:environmentd::serve:bootstrap: mz_persist_client::internal::encoding: halting process: 0.75.3 received persist state from the future 0.76.0 

This sounds like you might be downgrading?

**Comment by def-** [comment]
>  default/environmentd-0[environmentd]: error: Found argument '--timestamp-oracle-url' which wasn't expected, or isn't valid in this context 

And this sounds like an older version is being called with new `--timestamp-oracle-url` parameters.

**Comment by nrainer-materialize** [comment]
Should be fixed now, let's see what Nightly says.

File: misc/python/materialize/version_list.py [hunk 1]
```diff
@@ -222,73 +220,3 @@ def get_previous_mz_version(
...
-    MzVersion(major=0, minor=27, patch=0, prerelease=None, build=None)
-    """
-
-    def __init__(self) -> None:
-        files = Path(MZ_ROOT / "doc" / "user" / "content" / "releases").glob("v*.md")
-        self.versions = []
-        for f in files:
-            base = f.stem
-            metadata = frontmatter.load(f)
-            if not metadata.get("released", False):
```
Comment (nrainer-materialize) [inline misc/python/materialize/version_list.py:118]: Did we exclude not released versions here on purpose? Was the goal to skip older not released versions or to wait for the (latest) version to reach production?

**Comment by nrainer-materialize** [comment]
> > default/environmentd-0[environmentd]: error: Found argument '--timestamp-oracle-url' which wasn't expected, or isn't valid in this context
> 
> And this sounds like an older version is being called with new `--timestamp-oracle-url` parameters.

My assumption is that we see here an error which we don't see yet with the previous logic. This is because the previous logic waited for a version to appear in `doc/user/content/releases` with `released: true`. The test runs on main with `v0.77.1 ` as latest released version and succeeds. In this PR, the test is fed with ` v0.78.13`.

**Comment by philip-stoev** [comment]
I defer to @def on this one.

File: misc/python/materialize/version_list.py (hunk 1 above)
Comment (nrainer-materialize) [reply misc/python/materialize/version_list.py:118]: Ok, thanks. I understand that it is okay to change this behavior?

**Comment by nrainer-materialize** [comment]
@def-: Could you help me with this error? "Platform checks upgrade in Cloudtest/K8s" [is failing.](https://buildkite.com/materialize/nightlies/builds/5418#018c1c3a-9d34-449e-9b5c-6d95f9e6aa65)

The problem is that

```
Setting ca-crt to contents of /share/secrets/ca.crt...
^^^ +++
2:1: error: reading /share/secrets/ca.crt: No such file or directory (os error 2)
```
triggered from `$ set-from-file ca-crt=/share/secrets/ca.crt` in `misc/python/materialize/checks/all_checks/kafka_protocols.py` fails.

The check `KafkaProtocols` is guarded by a condition requiring at least version `v0.78.0-dev`. This is why it is executed on my branch, which picks more recent versions than the docs provided.

Any suggestions on how to fix that? Do I need to copy that file somewhere?

**Comment by def-** [comment]
Disabled now ^

**Comment by nrainer-materialize** [comment]
> Disabled now ^

Thank you!

**Review by def- — APPROVED** [review]

**Comment by nrainer-materialize** [comment]
(Just reordered and squashed some commits.)

**Pull request merged**